### PR TITLE
[performance] Inline ReactChildReconciler.instantiateChildren

### PR DESCRIPTION
**Pull request by mridgway**
Follow up from https://github.com/facebook/react/pull/3717

### CODE REVIEW

File: src/renderers/shared/reconciler/ReactChildReconciler.js [hunk 1]
```diff
@@ -17,6 +17,28 @@ var ReactReconciler = require('ReactReconciler');
...
+
+function instantiateChildIntoContext(traverseContext, child, name) {
+  if (child == null) {
+    return;
+  }
+  if (traverseContext[name] !== undefined) {
+    if (__DEV__) {
+      warning(
+        true,
+        'mountChildren(...): Encountered two children with the same key, ' +
```
Comment (spicyj) [inline src/renderers/shared/reconciler/ReactChildReconciler.js]: (instantiateChildren)

File: src/renderers/shared/reconciler/ReactChildReconciler.js [hunk 2]
```diff
@@ -17,6 +17,28 @@ var ReactReconciler = require('ReactReconciler');
...
+        '`%s`. Child keys must be unique; when two children share a key, only ' +
+        'the first child will be used.',
+        name
+      );
+    }
+    return;
+  }
+
+  traverseContext[name] = instantiateReactComponent(child);
```
Comment (spicyj) [inline src/renderers/shared/reconciler/ReactChildReconciler.js]: Can you keep the `, null` here?

**Comment by spicyj** [comment]
Code looks fine. Will try to replicate your perf-test results and if it looks positive I'll merge.

**Comment by mridgway** [comment]
Updated and rebased.

**Comment by spicyj** [comment]
@jimfb Assigning to you.

**Comment by jimfb** [comment]
Yep, this diff is freakishly similar to my diff.  We decided to merge #4408, since I was already iterating on it with Sebastian.  For this reason, I'm closing out this diff.

@mridgway Your diff was spot on.  I truly wish we had seen it earlier.  If we had merged this back in May, it would have saved us a whole bunch of time/energy :P.  Ah well, such is life.  We didn't know then what we know now (didn't have the perf tests on our end).  If you have any other ideas for improving perf and/or otherwise improving React, please do ping us; we can do a video chat or grab lunch or whatever.

**Comment by spicyj** [comment]
@mridgway Sorry for dropping the ball on this one. Got caught up in preparing for the conference and then I was out on vacation for a couple weeks. I could've passed this to @jimfb when he started focusing on perf work but didn't think of it at the time. Please do let us know if you find more improvements and hopefully we can get them merged sooner next time.

**Comment by mridgway** [comment]
No worries. I'm just glad someone is looking into performance :+1:. 

@jimfb I may take you up on lunch sometime. I'm eager to work with you guys on optimizing the server rendering flow but need some more information on what the plan is before I can contribute. Looks like https://github.com/facebook/react/issues/4012 has some more details that I'll look into.

**Pull request closed without merging**